File: site/js/passphrase_gate.js
// Passphrase gate — Python twin: tools/passphrase_gate.py.
//
// SOFT gate, NOT encryption: this file ships in the public SPA and the gated
// content is public on GitHub regardless. PBKDF2 only raises the cost of
// brute-forcing the phrase from the (public) hash; it does not protect content.
//
// Single source: loaded by the browser via <script src="js/passphrase_gate.js">
// in site/index.html AND required by tests/test_passphrase_gate.js. The hashing
// stays byte-identical to the .py twin (shared vectors in
// tests/fixtures/passphrase_gate_vectors.json).
//
// GATE_SALT / GATE_ITERATIONS are the single source of truth for the KDF params
// (deploy-pages.yml greps them out to compute the injected hash).

(function () {
  var GATE_SALT = 'a3f1c92e6b4d80571e2c9f3a6d8b0e47'; // 16-byte hex (public)
  var GATE_ITERATIONS = 200000;
  var STORAGE_KEY = 'sy_gate';

  function hexToBytes(hex) {
    var out = new Uint8Array(hex.length / 2);
    for (var i = 0; i < out.length; i++) out[i] = parseInt(hex.substr(i * 2, 2), 16);
    return out;
  }
  function bytesToHex(buf) {
    var b = new Uint8Array(buf), s = '';
    for (var i = 0; i < b.length; i++) s += (b[i] < 16 ? '0' : '') + b[i].toString(16);
    return s;
  }

  function derivePassphraseHash(phrase) {
    // Node (tests): synchronous PBKDF2, wrapped so the API matches the browser's
    // async Web Crypto path. Both are RFC 2898 PBKDF2-HMAC-SHA256 -> identical.
    if (typeof module !== 'undefined' && module.exports) {
      var nodeCrypto = require('crypto');
      var hex = nodeCrypto
        .pbkdf2Sync(Buffer.from(phrase, 'utf-8'), Buffer.from(GATE_SALT, 'hex'), GATE_ITERATIONS, 32, 'sha256')
        .toString('hex');
      return Promise.resolve(hex);
    }
    var enc = new TextEncoder();
    return crypto.subtle
      .importKey('raw', enc.encode(phrase), { name: 'PBKDF2' }, false, ['deriveBits'])
      .then(function (key) {
        return crypto.subtle.deriveBits(
          { name: 'PBKDF2', salt: hexToBytes(GATE_SALT), iterations: GATE_ITERATIONS, hash: 'SHA-256' },
          key,
          256
        );
      })
      .then(function (bits) { return bytesToHex(bits); });
  }

  function verifyPassphrase(phrase, expectedHash) {
    if (!expectedHash) return Promise.resolve(false);
    return derivePassphraseHash(phrase).then(function (h) { return h === expectedHash; });
  }

  function isGateEnabled(expectedHash) { return !!expectedHash; }

  function isUnlocked(expectedHash, storage) {
    if (!expectedHash) return false;
    try { return storage.getItem(STORAGE_KEY) === expectedHash; } catch (e) { return false; }
  }

  function recordUnlock(expectedHash, storage) {
    try { storage.setItem(STORAGE_KEY, expectedHash); } catch (e) {}
  }

  var api = {
    GATE_SALT: GATE_SALT,
    GATE_ITERATIONS: GATE_ITERATIONS,
    STORAGE_KEY: STORAGE_KEY,
    derivePassphraseHash: derivePassphraseHash,
    verifyPassphrase: verifyPassphrase,
    isGateEnabled: isGateEnabled,
    isUnlocked: isUnlocked,
    recordUnlock: recordUnlock,
  };

  if (typeof module !== 'undefined' && module.exports) {
    module.exports = api; // Node (tests)
  } else {
    this.PassphraseGate = api; // Browser: single namespace global
  }
}).call(typeof globalThis !== 'undefined' ? globalThis : this);

File: tools/passphrase_gate.py
"""Passphrase-gate hash — Python twin of site/js/passphrase_gate.js.

Used by .github/workflows/deploy-pages.yml to turn the GATE_PASSPHRASE secret
into the PBKDF2 hash injected into the SPA (var APP_GATE_HASH). The literal
passphrase is NEVER printed or committed; only the hash is emitted.

This is a SOFT gate, NOT encryption: the hash ships publicly in the deployed
site and the content behind the gate is public on GitHub regardless. The JS
twin (site/js/passphrase_gate.js) must produce identical output, enforced by
the shared vectors in tests/fixtures/passphrase_gate_vectors.json.
"""

from __future__ import annotations

import hashlib


def derive_hash(phrase: str, salt_hex: str, iterations: int) -> str:
    """PBKDF2-HMAC-SHA256(phrase, salt) -> 32-byte digest as lowercase hex."""
    dk = hashlib.pbkdf2_hmac("sha256", phrase.encode("utf-8"), bytes.fromhex(salt_hex), iterations, dklen=32)
    return dk.hex()


def main(argv: list[str] | None = None) -> None:
    import argparse

    parser = argparse.ArgumentParser(description="Compute the passphrase-gate hash.")
    sub = parser.add_subparsers(dest="cmd", required=True)
    h = sub.add_parser("hash", help="phrase -> PBKDF2 hex hash")
    h.add_argument("phrase")
    h.add_argument("--salt", required=True, help="salt as hex")
    h.add_argument("--iterations", type=int, required=True)
    args = parser.parse_args(argv)

    if args.cmd == "hash":
        print(derive_hash(args.phrase, args.salt, args.iterations))


if __name__ == "__main__":
    main()
